Update DestroyResourceWarhead to support ResourceType and amount parameters

abc013 opened the pull request:
Closes #17244 and removes the `TODO` in `DestroyResourceWarhead`.
This PR adds the possibility to
~- specify resources that are targeted by the Warhead (Testcase: Throw nuke on gems, no effect)~
- specify resource amount to remove per cell or in total (Testcase for `PerTile`: Nuke Truck, Testcase for `Total`: Nuke)

~The testcase also shows how much resource was removed in total when using the `Total` parameter to ensure that about the correct amount gets removed (Testcase: Nuke).~

abc013 committed:
Update DestroyResourceWarhead to support Resource type and amount options.

abcdefg30 commented on OpenRA.Mods.Common/Warheads/DestroyResourceWarhead.cs:
```diff
@@ -39,9 +52,44 @@ public override void DoImpact(Target target, WarheadArgs args)
+					if (ResourceAmount <= 0)
+						resLayer.Destroy(cell);
+					else if (ResourceRemovalType == ResourceRemoval.PerTile)
+						resLayer.Harvest(cell, ResourceAmount);
```
We should rename this, calling `Harvest` here sounds odd.

abc013 replied:
The problem is that this is used primarily for harvesters and returns the resource type of the harvested cell... I'm looking for names right now, but I can't find a fitting one. What about `Take` or `Remove`?

abcdefg30 replied:
`Take` sounds good to me.

abc013 replied:
Renamed.

abcdefg30 reviewed (commented):
I'm rather sceptical that defining a total (not per tile) amount of resources to be removed is actually useful for modding (and doesn't result in weird left-over patterns due to "rounding").

pchote commented:
> I'm rather sceptical that defining a total (not per tile) amount of resources to be removed is actually useful for modding (and doesn't result in weird left-over patterns due to "rounding").

I agree. Better IMO to drop the `ResourceRemovalType` and keep this per-cell (or maybe even the `ResourceAmount` idea completely? do we actually have a target use case for this?).

Note that #19090 rewrites a lot of the logic that this interacts with, so it would be best if this could be rebased on top of that.

abc013 commented:
Rebased, updated and removed the total resource removal. Sorry for the delay.

`ResourceAmount` does not have a specific target case in the default mods, but could be used for balancing (e.g. demo truck exploding on an ore field not creating a huge gap that can be used to deploy a MCV) or visual reasons and IMO is a nice addition for modders.
We can also just remove the `TODO` with this PR if noone likes the `ResourceAmount` idea.

Smittytron commented:
Tested and LGTM. Can remove testcase now.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/4ef1faecd699832d2182ff55f921e07d426a6afe
